rotkehlchen/chain/evm/decoding/pendle/decoder.py

```diff
@@ -1,122 +1,317 @@
+                event.notes = f'Receive {amount_1} {token_1.symbol} from depositing into Pendle'
+
+            elif (
+                    event.asset == token_0 and
+                    event.amount == amount_0 and
                     event.event_type == HistoryEventType.RECEIVE and
                     event.event_subtype == HistoryEventSubType.NONE
-                ):
-                    event.counterparty = CPT_PENDLE
-                    event.event_type = HistoryEventType.WITHDRAWAL
-                    event.event_subtype = HistoryEventSubType.REMOVE_ASSET
-                    event.notes = f'Withdraw {amount} PENDLE from vote escrow'
-                    break
-            else:
-                log.error(f'Could not find pendle unlock transfer for transaction {context.transaction}')  # noqa: E501
+            ):  # selling pt/yt
+                event.counterparty = CPT_PENDLE
+                event.event_type = HistoryEventType.WITHDRAWAL
+                event.event_subtype = HistoryEventSubType.REDEEM_WRAPPED
+                event.notes = f'Withdraw {event.amount} {event.asset.resolve_to_asset_with_symbol().symbol} from Pendle'  # noqa: E501
+
+            elif (
+                    event.amount == amount_1 and
+                    event.asset == token_1 and
+                    event.event_type == HistoryEventType.SPEND and
+                    event.event_subtype == HistoryEventSubType.NONE
+            ):  # selling pt/yt
+                event.counterparty = CPT_PENDLE
+                event.event_type = HistoryEventType.SPEND
+                event.event_subtype = HistoryEventSubType.RETURN_WRAPPED
+                event.notes = f'Return {event.amount} {token_1.symbol} to Pendle'
 
         return DEFAULT_DECODING_OUTPUT
 
+    def _decode_pendle_swap(self, context: DecoderContext) -> DecodingOutput:
+        if context.tx_log.topics[0] != SWAP_SINGLE_TOPIC:
+            return DEFAULT_DECODING_OUTPUT
+
+        # Determine swap type and retrieve received token details.
+        # Pendle only emits an event for the token being swapped,
+        # so we must check transaction logs to find the received token
+        token_out = (
+            self.evm_inquirer.native_token
+            if (addr := bytes_to_address(context.tx_log.topics[2])) == ZERO_ADDRESS
+            else self.base.get_or_create_evm_token(addr)
+        )
+        amount_out = asset_normalized_value(
+            amount=int.from_bytes(context.tx_log.data[:32]),
+            asset=token_out,
+        )
+        # no transaction was found that touches 1inch so we don't check for it(i.e swap_type=5).
+        if (swap_type := int.from_bytes(context.tx_log.topics[1])) not in (1, 2, 4):
+            return DEFAULT_DECODING_OUTPUT
+
+        for tx_log in context.all_logs:
+            if (  # odos
+                    swap_type == 2 and
+                    tx_log.topics[0] == ODOS_SWAP_TOPIC and
```

Comment:
maybe outside the for loop check

```
if swap_type == 1:
     expect_log_topic = KYBERSWAP_SWAPPED_TOPIC
elif swap_type == 2:
    expect_log_topic = ODOS_SWAP_TOPIC
...
```

this can be combined with the check that you do above. If I'm not mistaken this saves you a check per loop branch in each log